### DISCUSSION TITLE
Syncing ObservableList to label element

### DISCUSSION THREAD

**Discussion by nibinho** [zauberzeug/nicegui#1990] [Q&A] [answered]
### Question

Hi,

I was playing around with the observables module, which I think is a great toolkit in general. 
However, I found something strange when trying to sync a ui.label element to an ObservableList:

```python
from nicegui import events, observables, ui


def handle_change(e: events.ObservableChangeEventArguments):
    label.set_text(items)
    label.update()  # why is this needed?


items = observables.ObservableList(
    [1, 2, 3],
    on_change=handle_change,
)
ui.button("Add item", on_click=lambda: items.append(len(items) + 1))
ui.button("Remove item", on_click=items.pop)
ui.button("Clear items", on_click=items.clear)

ui.button("Set text", on_click=lambda: label.set_text(42))

label = ui.label(items)


ui.run()
```
Is it intended behavior that I need to call label.update() after label.set_text()? Why isn't it handled the same way as a "normal" .set_text() such as in the click event of the button? Nothing critical at all of course, I was just wondering why it behaves this way.

Thank you guys for building this framework, it has been amazing to work with in the last months and I am looking forward to what the future will bring for it! 🚀

**Comment by falkoschindler** [comment, marked as answer]
Thanks for the warm words, @nibinho! 😊

That's a very interesting problem you're facing here. But I guess I found the root cause:
When you call `label.set_text(items)`, the binding mechanism exits early because the text reference didn't change. It is still the `items` object.
https://github.com/zauberzeug/nicegui/blob/efe275585a499367534ee98b4bb0a7066ff0bf4b/nicegui/binding.py#L63-L64

Here is a more compact reproduction:
```py
data = [1, 2, 3]
label = ui.label(data)
ui.button('Add 4', on_click=lambda: (data.append(4), label.set_text(data)))
```

Actually this problem can't happen because `text` is supposed to be of type `str`. So if you call `set_text(str(items))` or `set_text(''.join(items))`, everything is fine because strings are immutable in Python.

On the other hand you don't even need to call `set_text(items)`, since the text _is_ already set to `items`. But in order to update the client, you need to call `update()`. So your example might look like this:
```py
items = observables.ObservableList([1, 2, 3], on_change=lambda: label.update())

ui.button("Add item", on_click=lambda: items.append(len(items) + 1))
ui.button("Remove item", on_click=items.pop)
ui.button("Clear items", on_click=items.clear)

label = ui.label(items)
```

**Reply by nibinho** [reply]
Got it, thanks a lot for your elaborated answer! 😊
